tiledb/sm/array_schema/enumeration.h

```diff
@@ -0,0 +1,229 @@
+/**
+ * @file enumeration.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class Enumeration.
+ */
+
+#ifndef TILEDB_ENUMERATION_H
+#define TILEDB_ENUMERATION_H
+
+#include <iostream>
+
+#include "tiledb/common/common.h"
+#include "tiledb/common/types/untyped_datum.h"
+#include "tiledb/sm/buffer/buffer.h"
+#include "tiledb/sm/enums/datatype.h"
+#include "tiledb/sm/filesystem/uri.h"
+#include "tiledb/storage_format/serialization/serializers.h"
+
+namespace tiledb::sm {
+
+/** Defines an array enumeration */
+class Enumeration {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /** No default constructor. Use the create factory method. */
+  Enumeration() = delete;
+
+  DISABLE_COPY(Enumeration);
+  DISABLE_MOVE(Enumeration);
+
+  /** Destructor. */
+  ~Enumeration() = default;
+
+  /* ********************************* */
+  /*             OPERATORS             */
+  /* ********************************* */
+
+  DISABLE_COPY_ASSIGN(Enumeration);
+  DISABLE_MOVE_ASSIGN(Enumeration);
+
+  /* ********************************* */
+  /*                 API               */
+  /* ********************************* */
+
+  /** Create a new Enumeration */
+  shared_ptr<Enumeration> static create(
```

Comment:
Why not make the constructor public?

Reply:
This is forcing folks to use `shared_ptr<Enumeration>` since that's all the other APIs will consume. Exposing the constructor means users have to know not to use it and only ever use the create factory for `shared_ptr<Enumeration>`.